### DISCUSSION TITLE
Batch / I don't get it

### DISCUSSION THREAD

**Discussion by norb-42** [O365/python-o365#1099] [Q&A]
Hello, 

I checked the documentation and the source but I do not understand how the attribute "batch" is working.

**My need :** 
I have to get all email on a specific date ( actually each day) and do some check on the messages and attachement.

**What i have done** 

First I created my query : 
```
query = mailbox.new_query()
query.on_attribute('received_date_time').greater_equal(datetime(2024, 3, 14))
query.chain('and').on_attribute('received_date_time').less_equal(datetime(2024, 3, 15))    
```

Then I request to get the messages and here the first thing I do not understand why batch is an int shouldn't it be a Bool saying that I want to request all messages based on the limit sequence ?
```
msgs = inbox.get_messages(limit=20,batch=20, query=query, download_attachments=True)
```

and then I loop on msg and print them, here the second point I do not understand if my batch is a collection how it continue the batch based on this code ?
```
for message in msgs:
    print (f'\t{Count},{message.sender.address.encode("utf-8")},{message.received}  ')

```

Of course the resulr is showing only 20 lines out of more than 1000

I also have tested with all on one line no knowing how th eiteration is really working in the class 
```
for message in inbox.get_messages(limit=20,batch=20, query=query, download_attachments=True):
    print (f'\t{Count},{message.sender.address.encode("utf-8")},{message.received}  ')
```

But I got the same 20 answers

I have seen the code speacking about "Pagination" but it look that my "get_message" is only sendinf back message and not pagination object

Yes i'm new to python and I probably do not get something but I hope that you can still help me .

Thanks

**Comment by norb-42** [comment]
After reading again into the forum the proper way to use batch seem to be placing limit=None 

eg : 
```
for message in inbox.get_messages(limit=None,batch=20, query=query, download_attachments=True):
    print (f'\t{Count},{message.sender.address.encode("utf-8")},{message.received}  ')
```

**Reply by alejcas** [reply]
You can use limit and batch together.

Limit will tell the total items retrieved. Batch just groups the items in api calls.

limit=100 batch=50 will make two api calls to the server each returning 50 items. 

MS Graph has a top limit of 999 items per request (or something like this... I don't remember). So in order to retrieve 20.000 items you need to batch them: limit=20.000 batch=500 (or something smaller than 999)
